### DISCUSSION TITLE
Array limits for vertices, colors & co

### DISCUSSION THREAD

**Discussion by rainergericke** [vsg-dev/VulkanSceneGraph#894] [Q&A] [answered]
Reworking our 3d code for vsg v1.0.8 I recognized the array limits are limited to uint32_t. Is there a technical reason for it?

**Comment by timoore** [comment]
The relevant limits on buffer sizes are limited to uint_32 in Vulkan. If
you want to "map" data from larger storage areas into vertex attributes,
then you could play games with the _data member.

**Comment by robertosfield** [comment]
As Tim says the structures that we map data to on the Vulkan size using
uint32_t so the vsg::Data class mirrors this.

THe limits didn't change in v1.0.8, ivsg::Data and subclasses have always
had these limits, the changes were just cosmetic w.r.t type.

**Comment by rainergericke** [comment]
I see. Our old code worked with bigger arrays too, but it is broken now. For a general solution in our software it seems to be better, to avoid to big arrays.

Thanks

**Reply by robertosfield** [reply]
I am confused as how things work before but not now.  The commit in v1.0.8 related vsg::Data/Array was a change from std::size_t to size_t for consistency purposes: b3c2939141affcae2929c2b7821925214781d6fd

What OS and compilers are you using?  Is std::size_t and size_t different sizes?

**Reply by robertosfield** [reply]
Could you provide details of the compile errors you are seeing?

**Reply by rainergericke** [reply]
A small test program shows:
Size_t is 8 Byte
std::Size_t is 8 Byte

I am using MacOS 13.5 and the clang++ compiler include with Xcode.

When I try to build our code with vsg v1.0.8 I get these errors:

<img width="468" alt="Bildschirmfoto 2023-08-02 um 10 37 39" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/31243285/81b25a0b-d7f7-47f0-a47f-215136b7573a">
for these code lines
<img width="578" alt="Bildschirmfoto 2023-08-02 um 10 38 18" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/31243285/7ef84336-28db-4075-a2df-56ca65c20d2b">
<img width="486" alt="Bildschirmfoto 2023-08-02 um 10 38 57" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/31243285/3dc290b2-601f-4673-9bdd-3a84b45b8f63">
<img width="819" alt="Bildschirmfoto 2023-08-02 um 10 39 38" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/31243285/6b33c3b6-1fb5-4c29-a020-9b770929da90">

To see how the code worked look at the code excerpt:
[code_excerpt.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/12238035/code_excerpt.txt)

**Reply by robertosfield** [reply]
These errors aren't to do with data size, but simply a refactor of vsg::ShaderSet/GraphicsPipelineConfigurator to be more flexible/extensible and to support multiple DescriptorSets.  I couldn't avoid breaking backwards compatibility when making these improvements to these classes.

Have a look at the changes between v1.0.7 and v1.0.8 to VSG and the recent changes to vsgXchange and vsgExamples as they should illustrate the changes I had to make to work with the new API.

**Comment by rainergericke** [comment, marked as answer]
Your changes are ok, I was able to give my solution a much easier structure. Would be great if e.g. the number of vertices in VertexDraw could be set to a greater value than 65535, but if not the world will keep turning.

**Reply by robertosfield** [reply]
I don't understand the limit to 65535, vsg::VertexDraw has always used uint32_t mapping to the underlying vkCmdDraw parameters, nothing has changed there:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/nodes/VertexDraw.h

**Reply by rainergericke** [reply]
We have to set the number of vertices in VertexDraw, it is defined as:
        uint32_t vertexCount = 0;
In VertexIndexDraw we have to set the numer of indices, defined as:
        uint32_t indexCount = 0;
If the array sizes are greater than 65535, only a part of the arrays con be used.

**Reply by robertosfield** [reply]
65535 is a ushort/uint16_t not uint32_t.

Where is the chain is the limit to 65535 coming from?

I am not aware of any changes in v1.0.8 thought would force a constraint to uin16_t, this isn't a change I would consciously make so if something has crept in somewhere it will need fixing.  At this point I don't have a lead to investigate so would appreciate any insights you can provide on what you think is introducing this limit.

**Comment by timoore** [comment]
> We have to set the number of vertices in VertexDraw, it is defined as: uint32_t vertexCount = 0; In VertexIndexDraw we have to set the numer of indices, defined as: uint32_t indexCount = 0; If the array sizes are greater than 65535, only a part of the arrays con be used.

You might be getting confused by behavior of a specific loader, but the indices supplied to VertexIndexDraw can be uint32_t.

**Reply by rainergericke** [reply]
Sorry I mixed something up! You are right.

**Reply by robertosfield** [reply]
So is v1.0.8 OK then?

**Reply by rainergericke** [reply]
Yes, it is!

**Reply by robertosfield** [reply]
I('ll drink to that :-)
